File: app/src/main/java/com/bopr/android/smailer/SmsTransport.java
package com.bopr.android.smailer;

import android.app.PendingIntent;
import android.content.Context;
import android.content.Intent;
import android.telephony.SmsManager;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import static android.app.PendingIntent.getBroadcast;
import static android.telephony.SmsManager.getDefault;

public class SmsTransport {

    private static final Logger log = LoggerFactory.getLogger("SmsTransport");

    public static final String ACTION_SMS_SENT = "sms_sent";
    public static final String ACTION_SMS_DELIVERED = "sms_delivered";

    private final Context context;

    public SmsTransport(Context context) {
        this.context = context;
    }

    public void send(String message, String phone) {
        SmsManager manager = getDefault();

        PendingIntent sentIntent = getBroadcast(context, 0, new Intent(ACTION_SMS_SENT), 0);
        PendingIntent deliveredIntent = getBroadcast(context, 0, new Intent(ACTION_SMS_DELIVERED), 0);
        for (String m : manager.divideMessage(message)) {
            manager.sendTextMessage(phone, null, m, sentIntent, deliveredIntent);

            log.debug("Sent SMS: " + message + " to " + phone);
        }
    }

}

File: app/src/main/kotlin/com/bopr/android/smailer/util/SmsTransport.kt
package com.bopr.android.smailer.util

import android.content.Context
import android.os.Build
import android.telephony.SmsManager

/**
 *  Mockable wrapper for [SmsManager] which is final.
 */
@Mockable
class SmsTransport(val context: Context) {

    fun sendMessage(phone: String?, message: String?) {
        context.smsManager.apply {
            sendMultipartTextMessage(phone, null, divideMessage(message), null, null)
        }
    }

    companion object {

        val Context.smsManager: SmsManager
            get() = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
                getSystemService(SmsManager::class.java)
            } else {
                @Suppress("DEPRECATION")
                SmsManager.getDefault()
            }

    }
}
